app/get_proc_arg.go

```diff
@@ -0,0 +1,231 @@
+            echo $proc
+          fi
+        fi
+      done
+    }
+    get_k3s_data_dir() {
+      proc=$(find_k3s_proc)
+      if [ "$proc" != "" ]; then
+        path=$(cat $proc//cmdline | tr '\000' '\n' |  grep 'data-dir' -A 1 | tail -n1)
+        if [ "$path" == "" ]; then
+          path=$(cat $proc//cmdline | tr '\000' '\n' |  grep '\^-d$' -A 1 | tail -n1)
+        fi
+        if [ "$path" == "" ]; then
+          echo "Cannot found arg data-dir in proc k3s. Will use default data-dir."
+          echo '/var/lib/rancher/k3s'
+        else
+          echo $path
+        fi
+      fi
+    }
+    get_k3s_data_dir
+	`
+)
+
+func getProcArg(kubeClient *clientset.Clientset, managerImage, name string) (string, error) {
+	switch name {
+	case ArgFlexvolumePluginDir:
+		dir, err := discoverFlexvolumeDir(kubeClient, managerImage)
+		if err != nil {
+			logrus.Errorf("Failed to get arg volume-plugin-dir, err: %v", err)
+			return "", err
+		}
+		return dir, nil
+	case ArgKubeletRootDir:
+		dir, err := discoverKubeletRootDir(kubeClient, managerImage)
+		if err != nil {
+			logrus.Errorf("Failed to get arg root-dir, err: %v", err)
+			return "", err
+		}
+		return dir, nil
+	}
+	return "", fmt.Errorf("Getting arg %v is not supported", name)
+}
+
+func discoverFlexvolumeDir(kubeClient *clientset.Clientset, managerImage string) (string, error) {
+	detectFlexVolumeScript := fmt.Sprintf(DetectKubeletArgScript, ArgFlexvolumePluginDir, DefaultFlexvolumeDir)
+	dir, err := discoverProcArg(kubeClient, managerImage, DetectFlexvolumeName, detectFlexVolumeScript)
+	if err != nil {
+		errors.Wrap(err, "Failed to get arg volume-plugin-dir in proc kubelet")
+		return "", err
+	}
+	if dir != "" {
+		return dir, nil
+	}
+	return "", errors.Wrap(fmt.Errorf("No related proc for volume-plugin-dir detection"), "Cannot get flexvolume dir, error out")
+}
+
+func discoverKubeletRootDir(kubeClient *clientset.Clientset, managerImage string) (string, error) {
+	// try to detect root-dir in proc kubelet
+	detectKubeletRootDirScript := fmt.Sprintf(DetectKubeletArgScript, ArgKubeletRootDir, DefaultKubeletRootDir)
+	rootDir, err := discoverProcArg(kubeClient, managerImage, DetectKubeletRootDirName, detectKubeletRootDirScript)
+	if err != nil {
+		errors.Wrap(err, "Failed to get arg root-dir in proc kubelet")
+		return "", err
+	}
+	// not found in proc kubelet. then try to detect root-dir in proc k3s
+	if rootDir == "" {
+		rootDir, err = discoverProcArg(kubeClient, managerImage, DetectKubeletRootDirName, DetectK3SDataDirScript)
+		// proc k3s exists.
+		if rootDir != "" {
+			rootDir = filepath.Join(rootDir, "/agent/kubelet")
+		}
+	}
+	if err != nil {
+		errors.Wrap(err, "Failed to get arg root-dir in proc k3s")
+		return "", err
+	}
+	if rootDir != "" {
+		return rootDir, nil
+	}
+	// nothing found. error out
+	return "", errors.Wrap(fmt.Errorf("No related proc for root-fir detection"), "Cannot get kubelet root dir, error out")
```

Comment:
Wrap is used for wrapping an existing error. So here you don't need it since you're generating the first error as well.

One pattern you can use is like this: https://github.com/rancher/longhorn-manager/blob/master/controller/volume_controller.go#L204

Also, normally the error should be in the branch. The expected result should be at last of the function.